aiida_vasp/parsers/vasp.py

```diff
@@ -17,78 +115,166 @@ class VaspParser(BaseParser):
+            'DOSCAR': None,
+            'IBZKPT': None,
+            'OUTCAR': None,
+            'EIGENVAL': None,
+        }
+
+        self._quantities_to_parse = []
+        self._output_nodes = {}
 
     def parse_with_retrieved(self, retrieved):
+
         self.check_state()
         self.out_folder = self.get_folder(retrieved)
+
         if not self.out_folder:
             return self.result(success=False)
-        outcar = self.get_file('OUTCAR')
-        if not outcar:
-            self.logger.error('OUTCAR not found, ' + 'look at the scheduler output for troubleshooting')
-            return self.result(success=False)
 
-        self.vrp = self.read_run()
+        # Get all specialised file parsers. Warnings will be issued if a file should be parsed and
+        # the corresponding files do not exist.
+        success = self._set_file_parsers()
 
-        bands, kpout, structure = self.read_eigenval()
+        if not success:
+            # A critical file i.e. OUTCAR does not exist. Abort parsing.
+            return self.result(success=False)
 
-        self.dcp = self.read_dos()
-        dosnode = self.get_dos_node(self.vrp, self.dcp)
+        # Get an initial list of quantities which should be parsed.
+        self._update_parsing_list()
 
-        if bands:
-            self.set_bands(bands)  # append output nodes
+        # Parse all implemented quantities in the nodesToAdd list, if they should be parsed. The list
+        # might get dynamically updated during the loop.
+        while self._quantities_to_parse:
+            quantity = self._quantities_to_parse.pop(0)
+            if self._settings['add_' + quantity]:
+                if not self._check_prerequesites(quantity):
+                    continue
+                self._output_nodes.update(getattr(self, '_get_' + quantity)())
 
-        if not kpout:
-            kpout = self.read_ibzkpt()
+        # Add output nodes if the corresponding data exists.
+        for key, value in self._output_nodes.iteritems():
+            if value:
+                self._set_node(key, value)
 
-        if kpout:
-            self.set_kpoints(kpout)
+        return self.result(success=True)
 
-        if structure:
-            self.set_structure(structure)
+    def _check_and_validate_settings(self):
+        """Check the settings and set which files should be parsed based on the input."""
 
-        if self.vrp:
-            if self.vrp.is_sc:  # add chgcar ouput node if selfconsistent run
-                chgnode = self.get_chgcar()
-                self.set_chgcar(chgnode)
+        import copy
+        new_settings = copy.deepcopy(self._settings)
 
-            if self.vrp.is_sc:
-                self.set_wavecar(self.get_wavecar())
+        for key, value in self._settings.iteritems():
+            if not key.startswith('add_'):
+                continue
+            quantity = key[4:]
+            for filename in PARSABLE_QUANTITIES[quantity]['parsers']:
```

Comment:
If the given quantity is not in PARSABLE_QUANTITIES this will give an unhelpful `KeyError`. Maybe we should do this check here and raise a more appropriate error with a more helpful error message? It might make sense to log a warning too.
In addition it might be good to check this in when verifying the inputs for the calculation but that would need some designing, as it should work for whichever parser is chosen...

Reply:
Valid point I forgot that there also might be bad user input. I was also thinking about checking these settings already when the calculation is setup. One could also modify the retrieve_list already at that point to retrieve the required files. My current answer to this question was to let these checks be done within a workchain when validating other inputs. But we could also add that to the calculation class.